# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/JSJitFrameIter.cpp
# Commit: 6eb32b942a04
# Full Hash: 6eb32b942a0497103ec39cf48708e0948c2d5565
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2024-07-29 20:43:57
# Regressor Bug: 1909224
# File Overlap Count: 5
# Description:
#   Bug 1909224 part 2 - Unify Wasm and JS exception handling code more. r=rhunt
#   
#   If there's no JS or Wasm catch handler on the stack, this patch lets us unwind
#   all the way to the entry frame in all cases, instead of returning to the JIT entry
#   stub and then calling the JS JIT exception handler from there.
# ==============================================================================

diff -r d8768f163b46 -r 6eb32b942a04 js/src/jit/JSJitFrameIter.cpp
--- a/js/src/jit/JSJitFrameIter.cpp	Mon Jul 29 13:14:47 2024 +0000
+++ b/js/src/jit/JSJitFrameIter.cpp	Mon Jul 29 13:14:47 2024 +0000
@@ -39,14 +39,18 @@
 }
 
 JSJitFrameIter::JSJitFrameIter(const JitActivation* activation,
-                               FrameType frameType, uint8_t* fp)
+                               FrameType frameType, uint8_t* fp, bool unwinding)
     : current_(fp), type_(frameType), activation_(activation) {
   // This constructor is only used when resuming iteration after iterating Wasm
   // frames in the same JitActivation so ignore activation_->bailoutData().
   //
   // Note: FrameType::JSJitToWasm is used for JIT => Wasm calls through the Wasm
   // JIT entry trampoline. FrameType::Exit is used for direct Ion => Wasm calls.
-  MOZ_ASSERT(fp > activation->jsOrWasmExitFP());
+  if (unwinding) {
+    MOZ_ASSERT(fp == activation->jsExitFP());
+  } else {
+    MOZ_ASSERT(fp > activation->jsOrWasmExitFP());
+  }
   MOZ_ASSERT(type_ == FrameType::JSJitToWasm || type_ == FrameType::Exit);
   MOZ_ASSERT(!TlsContext.get()->inUnsafeCallWithABI);
 }